src/beliefpropagation/beliefpropagation.jl

```diff
@@ -1,104 +1,67 @@
-  npartitions=nothing,
-  subgraph_vertices=nothing,
-  kwargs...,
+  ptn::PartitionedGraph; itensor_constructor=inds_e -> ITensor[dense(delta(inds_e))]
 )
-  return message_tensors(
-    partition(tn; nvertices_per_partition, npartitions, subgraph_vertices); kwargs...
-  )
-end
-
-function message_tensors_skeleton(subgraphs::DataGraph)
-  mts = DataGraph{vertextype(subgraphs),vertex_data_type(subgraphs),ITensorNetwork}(
-    directed_graph(underlying_graph(subgraphs))
-  )
-  for v in vertices(mts)
-    mts[v] = subgraphs[v]
-  end
-
-  return mts
-end
-
-function message_tensors(
-  subgraphs::DataGraph;
-  itensor_constructor=inds_e -> ITensor[dense(delta(i)) for i in inds_e],
-)
-  mts = message_tensors_skeleton(subgraphs)
-  for e in edges(subgraphs)
-    inds_e = commoninds(subgraphs[src(e)], subgraphs[dst(e)])
-    itensors = itensor_constructor(inds_e)
-    mts[e] = ITensorNetwork(itensors)
-    mts[reverse(e)] = dag(mts[e])
+  mts = Dict()
+  for e in partitionedges(ptn)
+    src_e_itn = subgraph(ptn, src(e))
+    dst_e_itn = subgraph(ptn, dst(e))
+    inds_e = commoninds(src_e_itn, dst_e_itn)
+    mts[e] = itensor_constructor(inds_e)
+    mts[reverse(e)] = dag.(mts[e])
   end
   return mts
 end
 
 """
-DO a single update of a message tensor using the current subgraph and the incoming mts
+Do a single update of a message tensor using the current subgraph and the incoming mts
 """
 function update_message_tensor(
-  tn::ITensorNetwork,
-  subgraph_vertices::Vector,
-  mts::Vector{ITensorNetwork};
-  contract_kwargs=(; alg="density_matrix", output_structure=path_graph_structure, maxdim=1),
+  ptn::PartitionedGraph, edge::PartitionEdge, mts; contract_kwargs=(; alg="exact")
 )
-  mts_itensors = reduce(vcat, ITensor.(mts); init=ITensor[])
+  pedges = setdiff(
+    partitionedges(ptn, boundary_edges(ptn, vertices(ptn, src(edge)); dir=:in)),
+    [reverse(edge)],
+  )
+  incoming_messages = [mts[e_in] for e_in in pedges]
+  incoming_messages = reduce(vcat, incoming_messages; init=ITensor[])
 
-  contract_list = ITensor[mts_itensors; ITensor[tn[v] for v in subgraph_vertices]]
-  tn = if isone(length(contract_list))
-    copy(only(contract_list))
-  else
-    ITensorNetwork(contract_list)
-  end
+  contract_list = ITensor[
+    incoming_messages
+    ITensor(subgraph(ptn, src(edge)))
+  ]
 
   if contract_kwargs.alg != "exact"
-    contract_output = contract(tn; contract_kwargs...)
-  else
-    contract_output = contract(tn; sequence=contraction_sequence(tn; alg="optimal"))
-  end
-
-  itn = if typeof(contract_output) == ITensor
-    ITensorNetwork(contract_output)
+    mt = first(contract(ITensorNetwork(contract_list); contract_kwargs...))
   else
-    first(contract_output)
+    mt = contract(
+      contract_list; sequence=contraction_sequence(contract_list; alg="optimal")
+    )
   end
-  normalize!.(vertex_data(itn))
 
-  return itn
-end
+  mt = isa(mt, ITensor) ? ITensor[mt] : ITensor(mt)
+  normalize!.(mt)
```

Comment:
This code doesn't feel right to me. Can't was always use `Vector{ITensor}`?

Also, I think it would be better to do out-of-place normalization, for example `mt .= normalize.(mt)`.

Reply:
Yeah I agree there is some way to unify this. Currently `contract` could return either an `ITensor` or a tuple corresponding to an `ITensorNetwork` and a `Float64` (from the return of `approx_ITensorNetwork`). So I was trying to account for both of these but maybe it should be some default. I could also imagine a world where `contract` returns `Vector{ITensor}` too

Reply:
I think one way to think about this is to not decide on a particular output of the message tensor update function, but rather require it to output the correct format/data for the message tensor cache.